src/shared/telemetry/awsTelemetryOptOut.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import { ext } from '../extensionGlobals'
+import { SettingsConfiguration } from '../settingsConfiguration'
+import { TelemetryService } from './telemetryService'
+const localize = nls.loadMessageBundle()
+
+export class AwsTelemetryOptOut {
+    private static readonly AWS_TELEMETRY_KEY = 'telemetry'
+    private static readonly TELEMETRY_OPT_OUT_SHOWN = 'awsTelemetryOptOutShown'
+
+    private readonly responseYes: string = localize('AWS.generic.response.yes', 'Yes')
+    private readonly responseNo: string = localize('AWS.generic.response.no', 'No')
+
+    public constructor(
+        public readonly service: TelemetryService,
+        private readonly settings: SettingsConfiguration
+    ) {
+        vscode.workspace.onDidChangeConfiguration(async event => {
+            // if telemetry.enableTelemetry changed and user has not expressed a preference
+            if (event.affectsConfiguration('telemetry.enableTelemetry')
+                && this.settings.readSetting(AwsTelemetryOptOut.AWS_TELEMETRY_KEY) === undefined
+            ) {
+                await this.updateTelemetryConfiguration(undefined)
+            }
+        })
+    }
+
+    public async updateTelemetryConfiguration(response: string | undefined) {
```

Comment:
I feel like this function should take an enum as its arg, to define the expected inputs. It does mean a conversion would be required from "localized button result" to suitable enum value though.

Reply:
:+1: